activation/activation.go

```diff
@@ -562,23 +562,26 @@ func (b *Builder) PublishActivationTx(ctx context.Context) error {
 func (b *Builder) createAtx(ctx context.Context) (*types.ActivationTx, error) {
 	b.log.With().Info("challenge ready")
 
-	pubEpoch := b.challenge.PublishEpoch()
-
 	hash, err := b.challenge.Hash()
 	if err != nil {
 		return nil, fmt.Errorf("getting challenge hash failed: %w", err)
 	}
 
 	// the following method waits for a PoET proof, which should take ~1 epoch
-	expireLayer := (pubEpoch + 2).FirstLayer()
-	atxExpired := b.layerClock.AwaitLayer(expireLayer) // this fires when the target epoch is over
+	pubEpoch := b.challenge.PublishEpoch()
+	// Add a 10% margin to the time it takes to generate a PoST proof
+	postDurationWithMargin := time.Duration(float64(b.lastPostGenDuration.Nanoseconds()) * 1.1)
+	postDurationInLayers := b.layerClock.DurationToLayers(postDurationWithMargin)
+	deadlineLayer := b.challenge.PubLayerID.Add(types.GetLayersPerEpoch()).Sub(postDurationInLayers)
+	deadline := b.layerClock.AwaitLayer(deadlineLayer)
```

Comment:
imo `AwaitLayer` should take a context and return a new context that is canceled when the layer is reached.

At the moment there are two signals that every `select` has to wait for: `<-ctx.Done()` and `<-deadlineChan`. Both of which serve the same purpose -> sending a signal to cancel when the result of `BuildNIPost` isn't needed any more.

I think with some minor changes to `layerClock` this could return a context that can then be used by `BuildNIPost` without the need of a second channel.

Reply:
Discussion continued in https://github.com/spacemeshos/go-spacemesh/pull/3633/